### ISSUE TITLE
usless_cast warnings with gcc 9.3 and 10.1 (C++17)

### ISSUE THREAD

**Issue by madeso** [nlohmann/json#2114]
### What is the issue you have?

gcc (g++) gives 2 usless_cast warnings when including json.hpp

### Please describe the steps to reproduce the issue. Can you provide a small but working code example?

```cpp
#include "json.hpp"

int main()
{
    return 0;
}
```
` g++ -Wuseless-cast -std=c++17 main.cc`

### What is the expected behavior?
No warnings.

### And what is the actual behavior instead?

2 compiler warnings:

```
In file included from main.cc:1:
json.hpp: In member function ‘bool nlohmann::detail::binary_reader<BasicJsonType, SAX>::parse_cbor_internal(bool)’:
json.hpp:5691:81: warning: useless cast to type ‘std::size_t’ {aka ‘long unsigned int’} [-Wuseless-cast]
 5691 |                 return get_number(input_format_t::cbor, len) and get_cbor_array(static_cast<std::size_t>(len));
      |                                                                                 ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~
json.hpp:5745:82: warning: useless cast to type ‘std::size_t’ {aka ‘long unsigned int’} [-Wuseless-cast]
 5745 |                 return get_number(input_format_t::cbor, len) and get_cbor_object(static_cast<std::size_t>(len));
      |                                                                                  ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

### Which compiler and operating system are you using? Is it a [supported compiler](https://github.com/nlohmann/json#supported-compilers)?

*supported*: g++ 9.3.0 (Ubuntu 9.3.0-11ubuntu0~18.04.1) 
*not supported but might be working*: g++ 10.1.0 (arch linux)

### Did you use a released version of the library or the version from the `develop` branch?

For g++ 9.3.0, only latest release [v3.7.3](https://github.com/nlohmann/json/releases/tag/v3.7.3) was tested
For g++ 10.1.0 both 3.7.3 and develop branch can reproduce it

### If you experience a compilation error: can you [compile and run the unit tests](https://github.com/nlohmann/json#execute-unit-tests)?

Not a compilation error, but a warning

**Comment by nlohmann** [comment]
Hm, this is very pedantic. In both cases, we cast a `std::uint64_t` to `std::size_t`. These types *may* be the same, but they do not need to be. So I would assume this warning is a false positive.

**Comment by madeso** [comment]
Hrm... you might be right about the false-positive.

I have a similar (open) issue in spdlog and it seems that *that* useless-cast can be removed by storing the offending code in a auto variable and then static_casting, making me considering removing the useless-cast warning from my compiler settings altogether.

**Comment by nlohmann** [comment]
In this specific case, the types are important as they are defined by CBOR. So I think there is nothing I can do.

**Comment by nlohmann** [comment]
Fixed with #2902.
